esti/s3_gateway_test.go

```diff
@@ -144,90 +144,173 @@ func TestS3ReadObject(t *testing.T) {
 	})
 
-	t.Run("HeadObject", func(t *testing.T) {
-		t.Run("Exists", func(t *testing.T) {
-			info, err := client.StatObject(ctx, repo, goodPath, minio.StatObjectOptions{})
-			if err != nil {
-				t.Errorf("client.StatObject(%s, %s): %s", repo, goodPath, err)
-			}
-			verifyObjectInfo(t, info, len(contents))
+	t.Run("get_no_physical_object", func(t *testing.T) {
+		blockStoreType := viper.GetString("blockstore_type")
+		if blockStoreType != "s3" {
+			t.Skip("Skipping test - blockstore type is not s3")
+		}
+		awsAccessKeyID := viper.GetString("aws_access_key_id")
+		awsSecretAccessKey := viper.GetString("aws_secret_access_key")
+		if awsAccessKeyID == "" || awsSecretAccessKey == "" {
+			t.Skip("Skipping test - AWS credentials not set")
+		}
+
+		// prepare entry with no physical object under it
+		const (
+			objPath   = "no-physical-object"
+			objS3Path = "main/" + objPath
+		)
+		_, err = minioClient.PutObject(ctx, repo, objS3Path, strings.NewReader(""), int64(0), minio.PutObjectOptions{})
+		if err != nil {
+			t.Errorf("PutObject(%s, %s): %s", repo, objS3Path, err)
+		}
+		statResp, err := client.StatObjectWithResponse(ctx, repo, "main", &apigen.StatObjectParams{
+			Path:    objPath,
+			Presign: swag.Bool(false),
 		})
-		t.Run("Doesn't exist", func(t *testing.T) {
-			info, err := client.StatObject(ctx, repo, badPath, minio.StatObjectOptions{})
-			if err == nil {
-				t.Errorf("client.StatObject(%s, %s): expected an error but got %+v", repo, badPath, info)
-			}
+		if err != nil {
+			t.Fatalf("StatObject(%s, %s): %s", repo, objPath, err)
+		}
+		if statResp.JSON200 == nil {
+			t.Fatalf("StatObject(%s, %s): got nil response, status: %s", repo, objPath, statResp.Status())
+		}
+		physicalAddress := statResp.JSON200.PhysicalAddress
 
-			s3ErrorResponse := minio.ToErrorResponse(err)
-			if s3ErrorResponse.StatusCode != 404 {
-				t.Errorf("Got %+v [%d] on reading when expecting Not Found [404]",
-					s3ErrorResponse, s3ErrorResponse.StatusCode)
-			}
+		// setup s3 client and delete the physical object
+		cfg, err := config.LoadDefaultConfig(ctx, config.WithCredentialsProvider(
+			awscredentials.NewStaticCredentialsProvider(awsAccessKeyID, awsSecretAccessKey, "")),
+		)
+		if err != nil {
+			t.Fatalf("failed to load aws configuration, %v", err)
+		}
+		s3Client := s3.NewFromConfig(cfg)
+
+		s3PhysicalURL, err := url.Parse(physicalAddress)
+		if err != nil {
+			t.Fatalf("failed to parse physical address %s, %v", physicalAddress, err)
+		}
+		if s3PhysicalURL.Scheme != "s3" {
+			t.Fatalf("physical address %s is not an s3 address", physicalAddress)
+		}
+		s3PhysicalKey := strings.TrimLeft(s3PhysicalURL.Path, "/")
+		_, err = s3Client.DeleteObject(ctx, &s3.DeleteObjectInput{
+			Bucket: aws.String(s3PhysicalURL.Host),
+			Key:    aws.String(s3PhysicalKey),
 		})
+		if err != nil {
+			t.Fatalf("DeleteObjects(%s, %s): %s", repo, objPath, err)
+		}
+
+		// wait for object to be gone
+		waiter := s3.NewObjectNotExistsWaiter(s3Client)
```

Comment:
Why is this waiter still needed?  I thought S3 was now list-after-write consistent...?  (Is MinIO _not_ consistent for this?)

Reply:
Good point - I was using our current Delete implementation in the block adapter.
Fixing and open an issue for block adapter.

Reply:
https://github.com/treeverse/lakeFS/issues/6836